### DISCUSSION TITLE
Feedback on suitable architecture

### DISCUSSION THREAD

**Discussion by GianlucaDeStefano** [apache/airflow#23004] [Q&A] [answered]
Good day, I am  currently setting up a pipeline that has to crawl and process data from hundreds of thousands of urls.
In short: for each predefined webpage every day I have to download its content to see if it has changed from previous days.

This seems easy enough but there are some peculiarities that I don't know how to handle since I am new to airflow. 
I have two possible architectures in my mind at the moment:

1) First architecture: Create a DAG to process an individual url and trigger it once for each link every day. 
The problem is that  since I have huge amount of links to process, this DAG would be triggered a huge amount of times every single day and I don't think this is the use-case DAGS where intended for. On the other hand however this architecture would allow me to use all the neat features that airflow provides(i.e.: re-runs).   

2) In the second architecture the idea is to have a DAG with just 2 tasks: 
      1. Load -> Load all urls to process inside a queue
      2. Execute -> spawn some workers to process the urls in the queue. 
      This dag would be executed once per day to  process all the urls. The problem of this approach is that I would lose too much control over the runs of individual urls and furthermore I could not make use of all the features of airflow (i.e.: re-runs in case of failures of the pipeline on individual links).

My question is therefore the following: what's the best path to follow to maintain flexibility and granularity in the management of the pipeline? Is there a third way I am not seeing? 
Thank you in advance.

**Comment by potiuk** [comment, marked as answer]
depends how big the tasks are (each task is an overhead for scheduling and executoin) but likely your best choice will be something that is coming in 2.3.0 - Dynamic task mapping: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-42+Dynamic+Task+Mapping

You will be able to define a single task - parameterized with the array of urls and fan-out-> fan-in tasks - have one task for each URL (and you can define max parallelism etc.).

If each parsing/processing takes a few seconds, then the overhead with Celery Executor might not be big, and indeed - the ability to re-run stuff that failed will be nice.

**Reply by potiuk** [reply]
BTW. 2.3.0 is going to be released in a few weeks I think (depends on the progress of testing).

**Reply by GianlucaDeStefano** [reply]
Thanks! A single task will be very light. I think it will contain just an http request plus some other transformation. 

I see thanks! Unfortunately however I can't wait for these features as the pipeline has to be up and running ASAP but I will keep them in mind and update it in the future. 

I think that I will adopt the second approach and use airflow to trigger the macro load and execution operations.

**Reply by potiuk** [reply]
Yeah. If they are lightweight that makes more sense. Runing Every task in Airlfow involves some database queries, scheduling, passing tasks through several levels of abstractions, distributing to one of the free workers etc. So if your tasks are really  lightweigh and a lot of them, this might add a loooooot of overhead.
